src/2024h2/higher-level-rust.md

```diff
@@ -0,0 +1,283 @@
+
+Generally, a "high level Rust" would be fast-to-compile and maximally performant by default. The tweaks here do not require language changes and are generally a question of engineering effort rather than design consensus.
+
+
+### The "shiny future" we are working towards
+
+A "high level Rust" would be a Rust that has a strong focus on iteration speed. Developers would benefit from Rust's performance, safety, and reliability guarantees without the current status quo of long compile times, verbose code, and program architecture limitations.
+
+A "high level" Rust would:
+- Compile quickly, even for fresh builds
+- Be terse in the common case
+- Produce performant programs even in debug mode
+- Provide language shortcuts to get to running code faster
+
+In our "shiny future," an aspiring genomics researcher would:
+- be able to quickly jump into a new project
+- add powerful dependencies with little compile-time cost
+- use various procedural macros with little compile-time cost
+- cleanly migrate their existing program architecture to Rust with few lifetime issues
+- employ various shortcuts like unwrap to get to running code quicker
+
+----
+
+To imagine a fictional scenario, let's take the case of "Alex."
+
+Alex is a genomics researcher studying ligand receptor interaction to improve drug therapy for cancer. They work with very large datasets and need to design new algorithms to process genomics data and simulate drug interactions. Alex recently heard that Rust has a great genomics library (Genomix) and decides to try out Rust for their next project.
+
+Alex creates a new project and starts adding various libraries. To start, they add Polars and Genomix. They also realize they want to wrap their research code in a web frontend and allow remote data, so they add Tokio, Reqwest, Axum, and Dioxus. They write a simple program that fetches data from their research lab's S3 bucket, downloads it, cleans it, processes, and then displays it.
+
+For the first time, they type `cargo run.` The project builds in 10 seconds and their code starts running. The analysis completes churns for a bit and Alex is greeted with a basic webpage visualizing their results. They start working on the visualization interface, adding interactivity with new callbacks and async routines. Thanks to hotreloading, the webpage updates without fully recompiling and losing program state.
```

Comment:
I am very surprised by this number. For me adding the mentioned libraries (except genomix, as it doesn't look to be a thing) to an empty project and compiling from scratch took 1.5 minutes. Is there any "proof" or ideas on how this roughly 9x speedup is going to be achieved, or is this just a dream/wish?

Reply:
> Is there any "proof" or ideas on how this roughly 9x speedup is going to be achieved,

I can imagine some kind of sdk-style bundle layer in the middle.  While first-time building performance might not be improved drastically, the first-time building process will build and "deploy" the bundle on the machine. This bundle might include the mentioned "Tokio, Reqwest, Axum, and Dioxus" (and other large stuff like `wgpu`, `skia`, `v8`, etc...) packages, and generate a (`cdylib` + `rlib`) pair.

For development purpose, the end-product will link to the `rlib` and `cdylib` and the `cdylib` will never be rebuilt unless requested upon explicitly. With this approach, if the user code is magnitude less than the bundle implementation, the build and link time can reach 9x speedup.